src/theory/arith/equality_solver.h

```diff
@@ -0,0 +1,115 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Arithmetic equality solver
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__ARITH__EQUALITY_SOLVER_H
+#define CVC5__THEORY__ARITH__EQUALITY_SOLVER_H
+
+#include "context/cdhashset.h"
+#include "expr/node.h"
+#include "proof/trust_node.h"
+#include "theory/arith/arith_state.h"
+#include "theory/ee_setup_info.h"
+#include "theory/uf/equality_engine.h"
+
+namespace cvc5 {
+namespace theory {
+namespace arith {
+
+class InferenceManager;
+
+/**
+ * The arithmetic equality solver. This class manages arithmetic equalities
+ * in the default way via an equality engine.
+ *
+ * Since arithmetic has multiple ways of propagating literals, it tracks
+ * the literals that it propagates and only explains the literals that
+ * originated from this class.
+ */
+class EqualitySolver
+{
+  using NodeSet = context::CDHashSet<Node>;
+
+ public:
+  EqualitySolver(ArithState& astate, InferenceManager& aim);
+  ~EqualitySolver() {}
+  //--------------------------------- initialization
+  /**
+   * Returns true if we need an equality engine, see
+   * Theory::needsEqualityEngine.
+   */
+  bool needsEqualityEngine(EeSetupInfo& esi);
+  /**
+   * Finish initialize
+   */
+  void finishInit();
+  //--------------------------------- end initialization
+  /**
+   * Pre-notify fact, return true if we are finished processed, false if
+   * we wish to assert the fact to the equality engine of this class.
+   */
+  bool preNotifyFact(
+      TNode atom, bool pol, TNode fact, bool isPrereg, bool isInternal);
+  /**
+   * Return an explanation for the literal lit (which was previously propagated
+   * by this solver).
+   */
+  TrustNode explain(TNode lit);
+
+ private:
+  /** Notification class from the equality engine */
+  class EqualitySolverNotify : public eq::EqualityEngineNotify
```

Comment:
I know that this is kind of a common pattern that we use but why not simply have `EqualitySolver` inherit from `eq::EqualityEngineNotify` and get rid of this class?

(This is arguably a more general design decision and probably outside of the scope of this PR)

Reply:
I tend not to like doing this, because it is misleading to what the purpose of the class is.  Although that is one of its responsibilities, it's main purpose is to setup / manage the equality engine.  One can also get into annoying cases where you are tempted to use multiple inheritence, say e.g. if EqualitySolver wanted to inherit from some kind of Subsolver class.

Nevertheless it's an interesting thought.